src/tree/context-dep.cc

```diff
@@ -178,9 +178,105 @@ void ContextDependency::Read (std::istream &is, bool binary) {
+  vec.push_back(std::make_pair(kPdfClass, static_cast<EventValueType>(forward_pdf_class)));
+  std::sort(vec.begin(), vec.end());
+  to_pdf_->MultiMap(vec, &forward_pdfs);
+  SortAndUniq(&forward_pdfs);
+
+  // get list of possible self-loop pdfs
+  vec.clear();
+  for (size_t i = 0; i < N_; i++)
+    if (phone_window[i] > 0)
+      vec.push_back(std::make_pair(static_cast<EventKeyType>(i),
+                                   static_cast<EventValueType>(phone_window[i])));
+  vec.push_back(std::make_pair(kPdfClass, static_cast<EventValueType>(self_loop_pdf_class)));
+  std::sort(vec.begin(), vec.end());
+  to_pdf_->MultiMap(vec, &self_loop_pdfs);
+  SortAndUniq(&self_loop_pdfs);
+
+  if (forward_pdfs.size() == 1 || self_loop_pdfs.size() == 1) {
+    for (size_t m = 0; m < forward_pdfs.size(); m++)
+      for (size_t n = 0; n < self_loop_pdfs.size(); n++)
+        pairs->insert(std::make_pair(forward_pdfs[m], self_loop_pdfs[n]));
+  } else {
+    // Choose 'position' as a phone position in 'context' that's currently
+    // 0, and that is as close as possible to the central position P.
+    int32 min_dist = N_ + 1;
+    for (int32 i = 0; i < N_; i++) {
+      int32 dist = (P_ - i > 0) ? (P_ - i) : (i - P_);
+      if (phone_window[i] == 0 && dist < min_dist) {
+        position = i;
+        min_dist = dist;
+      }
+    }
+
+    for (size_t i = 0 ; i < phones.size(); i++) {
+      new_phone_window[position] = phones[i];
+      EnumeratePairs(phones, self_loop_pdf_class, forward_pdf_class,
+                     new_phone_window, pairs);
+    }
+  }
+}
+
+void ContextDependency::GetPdfInfo(
+    const std::vector<int32> &phones,
+    const std::vector<std::vector<std::pair<int32, int32> > > &pdf_class_pairs,
+    std::vector<std::vector<std::vector<std::pair<int32, int32> > > > *pdf_info) const {
+
+  KALDI_ASSERT(pdf_info != NULL);
+  pdf_info->resize(1 + *std::max_element(phones.begin(), phones.end()));
+  std::vector<int32> phone_window(N_);
+  EventType vec;
+  for (size_t i = 0 ; i < phones.size(); i++) {
+    // loop over phones
+    int32 phone = phones[i];
+    (*pdf_info)[phone].resize(pdf_class_pairs[phone].size());
+    for (size_t j = 0; j < pdf_class_pairs[phone].size(); j++) {
+      // loop over pdf_class pairs
+      int32 pdf_class = pdf_class_pairs[phone][j].first,
+            self_loop_pdf_class = pdf_class_pairs[phone][j].second;
+      for (size_t win_start = 0; win_start < phone_window.size(); win_start++) {
+	if (win_start != P_)
+          phone_window[win_start] = 0;
+	else
+	  phone_window[win_start] = phone;
+      }
+      unordered_set<std::pair<int32, int32>, PairHasher<int32> > pairs;
+      EnumeratePairs(phones, self_loop_pdf_class, pdf_class, phone_window, &pairs);
+      unordered_set<std::pair<int32, int32>, PairHasher<int32> >::iterator iter = pairs.begin(),
+                           end = pairs.end();
+      for (; iter != end; ++iter)
+        (*pdf_info)[phone][j].push_back(*iter);
```

Comment:
Before exiting the function you might want to sort all of these output vectors.  That will make the order deterministic... might be useful in some circumstances for testing purposes.